links/src/commonMain/kotlin/com/splendo/kaluga/links/utils/decoders.kt

```diff
@@ -0,0 +1,83 @@
+/*
+ Copyright 2020 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.links.utils
+
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.AbstractDecoder
+import kotlinx.serialization.encoding.CompositeDecoder
+import kotlinx.serialization.modules.EmptySerializersModule
+import kotlinx.serialization.modules.SerializersModule
+
+fun <T> decodeFromList(list: List<Any>, deserializer: DeserializationStrategy<T>): T {
+    val decoder = LinksDecoder(ArrayDeque(list))
+    return decoder.decodeSerializableValue(deserializer)
+}
+
+class LinksDecoder(
+    internal val list: ArrayDeque<Any>,
+    var elementsCount: Int = 0
+) : AbstractDecoder() {
+    private var elementIndex = 0
+
+    override val serializersModule: SerializersModule = EmptySerializersModule
+
+    override fun decodeElementIndex(descriptor: SerialDescriptor): Int {
+        if (elementIndex == descriptor.elementsCount) return CompositeDecoder.DECODE_DONE
+        return elementIndex++
```

Comment:
Will it return first and then increment or increment first and then return?

Reply:
return and increment